### ISSUE TITLE
Implement IPv6 Support for the Web Server

### ISSUE THREAD

**Issue by 552020** [dantol29/webserver#194]
Currently, the server only supports IPv4, which limits connectivity options for clients using IPv6. The subject doesn't specify anything about it, neither the RFC. There are two possibilities to implement this:

1. **Dual-Stack Mode** (Preferred): Modify the existing IPv6 socket to also handle IPv4 connections using IPv4-mapped IPv6 addresses. This approach reduces complexity by not requiring separate sockets and listening threads for each protocol.
2. **Separate Sockets**: Maintain distinct sockets for IPv4 and IPv6. This approach might be simpler to manage from a network policy perspective but increases resource usage and complexity.

**Code Changes:**

To implement dual-stack mode, you need to adjust the socket creation and configuration. Here’s how we could modify the code:

```c++

#include <sys/socket.h>

#include <netinet/in.h>

#include <arpa/inet.h>

int create_dual_stack_socket() {

    int sockfd = socket(AF_INET6, SOCK_STREAM, 0);

    if (sockfd < 0) {

        perror("Failed to create socket");

        return -1;

    }

    // Set IPV6_V6ONLY to 0 to enable dual-stack mode

    int no = 0;

    if (setsockopt(sockfd, IPPROTO_IPV6, IPV6_V6ONLY, &no, sizeof(no)) < 0) {

        perror("Failed to set IPV6_V6ONLY option");

        close(sockfd);

        return -1;

    }

    struct sockaddr_in6 addr;

    memset(&addr, 0, sizeof(addr));

    addr.sin6_family = AF_INET6;

    addr.sin6_addr = in6addr_any;  // Listen on all interfaces

    addr.sin6_port = htons(80);

    if (bind(sockfd, (struct sockaddr *)&addr, sizeof(addr)) < 0) {

        perror("Bind failed");

        close(sockfd);

        return -1;

    }

    return sockfd;

}
```

**Example GET Requests:**

* **IPv4 GET Request:**

  ```
  
  GET / HTTP/1.1
  
  Host: 192.168.1.5
  
  ```
* **IPv6 GET Request:**

  ```
  
  GET / HTTP/1.1
  
  Host: [2001:db8::1]
  
  ```

Basically we need to specify `AF_INET6` instead of `AF_INET4` on Socket creation and then turn off IPV6ONLY with setsockopt if it's turned on.

**Comment by dantol29** [comment]
@552020 Is it implemented?

**Comment by 552020** [comment]
I hope so! It was so painful! We have support for IPv4 and we work with IPv6, so the other way around. We should maybe just add in the test multi_server some IPv6 address to demonstrate that it works. Maybe another issue? Or just closing it.
